contracts/oracle/uniswap/UniswapV3OracleWrapper.sol

```diff
@@ -0,0 +1,193 @@
+    require(_uniswapWrapper != address(0x0), "_uniswapWrapper cannot be null address");
+    
+    validatePoolLiquidity(_oracleConfig.uniswapPool, _oracleConfig.minPoolLiquidity);
+    validateTokensInPool(_oracleConfig.uniswapPool, _inputToken, _outputToken);
+    validateTwapPeriod(_oracleConfig.twapPeriod);
+    
+    pool = _oracleConfig.uniswapPool;
+    inputToken = _inputToken;
+    outputToken = _outputToken;
+    uniswapWrapper = IUniswapWrapper(_uniswapWrapper);   
+    oracleConfig = _oracleConfig;
+
+    addSupportForPool(_oracleConfig.uniswapPool, oracleConfig.twapPeriod, meanBlockTime);
+    emit TwapPeriodUpdate(_oracleConfig.uniswapPool, 0, oracleConfig.twapPeriod); 
+  }
+
+  /// @notice updates the oracle price
+  /// @dev no-op, Uniswap is updated automatically
+  function update() external view override {}
+
+  // ----------- Getters -----------
+
+  /// @notice Convenience getter for twapPeriod
+  function getTwapPeriod() external override view returns (uint32) {
+    return oracleConfig.twapPeriod;
+  }
+
+  /// @notice Read the oracle price
+  function read() external view override returns (Decimal.D256 memory, bool) {
+    validatePoolLiquidity(pool, oracleConfig.minPoolLiquidity);
+    
+    uint8 inputTokenDecimals = ERC20(inputToken).decimals();
+    uint8 outputTokenDecimals = ERC20(outputToken).decimals();
+
+    uint256 rawPrice = uniswapWrapper.calculatePrice(
+      pool, 
+      oracleConfig.twapPeriod,
+      inputToken,
+      outputToken,
+      inputTokenDecimals,
+      outputTokenDecimals,
+      oracleConfig.precision
+    );
+
+    bool valid = !paused();
+    
+    // Divide out the extra decimals that decimal library adds
+    Decimal.D256 memory value = Decimal.from(rawPrice).div(10**18);
+    return (value, valid);
+  }
+
+  /// @notice no-op, Uniswap V3 constantly updates the price
+  function isOutdated() external pure override returns (bool) {
+    return false;
+  }
+
+
+  // ----------- Internal -----------
+
+  /// @notice Validate that the pool liquidity is above a safe minimum threshold
+  function validatePoolLiquidity(address _pool, uint128 _minPoolLiquidity) internal view {
+    require(
+      IUniswapV3Pool(_pool).liquidity() >= _minPoolLiquidity,
+      "Pool has insufficient liquidity"
+    );
+  }
+
+  /// @notice Validate that the single Uniswap pool has reserves in both input and output tokens
+  /// @param _pool Uniswap pool address
+  /// @param _inputToken Input token for which a price is being quoted
+  /// @param _outputToken Output token
+  function validateTokensInPool(address _pool, address _inputToken, address _outputToken) internal view {
+    address uniswapToken0 = IUniswapV3Pool(_pool).token0();
+    address uniswapToken1 = IUniswapV3Pool(_pool).token1();
+    (address token0, address token1) = sortTokensAccordingToUniswap(_inputToken, _outputToken);
+  
+    require(
+      uniswapToken0 == token0 || uniswapToken1 == token1,
```

Comment:
should be && ?

Reply:
Very good spot. Fixed and will add tests to cover this